Add docs for hipInfo and related build and tests (#5903)
ROCm/rocm-systems: projects/hip/docs/how-to/hip_runtime_api/hipInfo.rst

.. meta::
   :description: hipInfo documentation
   :keywords: AMD, ROCm, HIP, rocminfo, GPU diagnostics, device properties

.. _hipInfo:

Using hipInfo
=============

> ``hipInfo`` is a diagnostic utility that displays information about AMD GPU devices available on your system. It queries and reports device properties, architectural features, memory information, and peer-to-peer capabilities through the HIP runtime API. ``hipInfo`` is similar to ``rocminfo``, though they are different as described in :ref:`hip-v-rocm-info`. For more information, see `rocminfo <https://rocm.docs.amd.com/projects/rocminfo/en/latest/index.html>`__.

Reviewer: ```suggestion
``hipInfo`` is a diagnostic utility that displays information about AMD GPU devices available on your system. It queries and reports device properties, architectural features, memory information, and peer-to-peer capabilities through the HIP runtime API. While ``hipInfo`` is similar to ``rocminfo``, the two tools differ in scope and usage, as described in :ref:`hip-v-rocm-info`. For more information, see `rocminfo <https://rocm.docs.amd.com/projects/rocminfo/en/latest/index.html>`__.
```

``hipInfo`` can be used for:

* Verifying GPU detection and configuration
* Evaluating device capabilities for HIP application development
* Troubleshooting GPU-related issues
* Gathering system information for bug reports or support requests

.. note::

>    ``hipInfo`` is installed as part of the `HIP SDK for Windows <https://rocm.docs.amd.com/projects/install-on-windows/en/latest/>`_. However, on Linux systems it is not installed with HIP, but can be separately built as described in `Building HIP tests <https://rocm.docs.amd.com/projects/HIP/en/latest/install/build.html#build-hip-tests>`_. 

Reviewer 2: I think this reference to HIP SDK for Windows is fine. 
However, I'm not sure of its status.

>    ``hipInfo`` is installed as part of the `HIP SDK for Windows <https://rocm.docs.amd.com/projects/install-on-windows/en/latest/>`_. However, on Linux systems it is not installed with HIP, but can be separately built as described in `Building HIP tests <https://rocm.docs.amd.com/projects/HIP/en/latest/install/build.html#build-hip-tests>`_. 

Reviewer 2: ```suggestion
   ``hipInfo`` is installed as part of the `HIP SDK for Windows <https://rocm.docs.amd.com/projects/install-on-windows/en/latest/>`_. However, on Linux systems, it's not installed with HIP, but can be separately built as described in `Building HIP tests <https://rocm.docs.amd.com/projects/HIP/en/latest/install/build.html#build-hip-tests>`_. 
```

hipInfo command
----------------

> The command requires no arguments and will automatically detect and report information for all available AMD GPU devices.

Reviewer 2: ```suggestion
The command requires no arguments and automatically detects and reports information for all available AMD GPU devices.
```

.. code-block:: bash

   $ ./hipInfo

Output Information
------------------

``hipInfo`` reports support for various GPU features, such as:

* Atomic operations (int32, int64, float)

> * Double precision floating point

Reviewer 2: ```suggestion
* Double-precision floating point
```

* Warp/wavefront operations (vote, ballot, shuffle)
* Thread fence operations
* 3D grid support
* Texture capabilities

The tool also displays the following categories of information for each detected GPU:

- Device Identification

   * **Name**: GPU device name
   * **pciBusID, pciDeviceID, pciDomainID**: PCI location identifiers

>    * **gcnArchName**: AMD GCN architecture name (for example: gfx906, gfx90a, gfx1100)

Reviewer 3: ```suggestion
   * **gcnArchName**: AMD architecture name (for example: gfx906, gfx90a, gfx1100)
```

GCN is what we had before RDNA and CDNA.

I'm also conflicted regarding the "architecture name". RDNA and CDNA are architecture names. The gfx identifiers are LLVM target IDs; gfx1100, gfx1101, gfx1102 and gfx1103 all share the RDNA3 architecture.

> - Compute Capabilities

Reviewer 3: That wording is unfortunate. "Compute capability" is a [competitor term](https://developer.nvidia.com/cuda/gpus) that might be misleading in this case.

Reviewer 3: What is the conclusion here?

Author: That is an Engineering discussion. The docs are simply reporting what is returned by the tool.

   * **multiProcessorCount**: Number of compute units
   * **maxThreadsPerMultiProcessor**: Maximum threads per compute unit
   * **maxThreadsPerBlock**: Maximum threads per work-group
   * **maxThreadsDim**: Maximum work-group dimensions (x, y, z)
   * **maxGridSize**: Maximum grid dimensions (x, y, z)

>    * **warpSize**: Wavefront size (typically 64 for AMD GPUs)

Reviewer 3: ```suggestion
   * **warpSize**: Wavefront size (32 for AMD Radeon GPUs and 64 for AMD Instinct GPUs)
```

HIP doesn't support a wavefront size of 64 for Radeon GPUs so we don't need the "typically" IMO.

Reviewer 3: And here?

>    * **warpSize**: Wavefront size (typically 64 for AMD GPUs)

Reviewer 2: this description seems odd - does it need to be clarified?

   * **regsPerBlock**: Registers available per block

>    * **major/minor**: Compute capability version

Reviewer 3: We don't have "compute capabilities" as defined by the competition; our LLVM target IDs are the equivalent. On my gfx1103 GPU the output is `major: 11` and `minor: 0`. My guess is that the `3` was dropped and the rest was split into this major/minor format.

- Other Properties

   * **cooperativeLaunch**: Cooperative group launch support
   * **cooperativeMultiDeviceLaunch**: Multi-device cooperative launch support
   * **concurrentKernels**: Concurrent kernel execution support
   * **isIntegrated**: Whether GPU is integrated or discrete
   * **peers/non-peers**: P2P (peer-to-peer) capable device relationships

Example Output
--------------

.. code-block:: text

   --------------------------------------------------------------------------------
   device#                           0
   Name:                             AMD Radeon GPU
   pciBusID:                         103
   pciDeviceID:                      0
   pciDomainID:                      0
   multiProcessorCount:              64
   maxThreadsPerMultiProcessor:      2560
   clockRate:                        1800 Mhz
   memoryClockRate:                  1000 Mhz
   memoryBusWidth:                   4096
   totalGlobalMem:                   31.98 GB
   sharedMemPerBlock:                64.00 KB
   warpSize:                         64
   major:                            9
   minor:                            0

>    gcnArchName:                      gfx906:sramecc+:xnack-

Reviewer 3: The relationship between LLVM target IDs and the major/minor schema is also shown here.

   memInfo.total:                    31.98 GB
   memInfo.free:                     31.96 GB (100%)

.. _hip-v-rocm-info:

> Comparing hipinfo and rocminfo

Reviewer 3: ```suggestion
Comparing hipInfo and rocminfo
```

------------------------------

While both ``hipInfo`` and ``rocminfo`` provide GPU device information, they serve different purposes:

**hipInfo**:

* Uses the HIP runtime API (``hipGetDeviceProperties``)
* Shows device properties from the perspective of HIP application programming
* Displays information relevant to HIP kernel development
* Reports current memory usage and runtime state
* Provides information in a developer-friendly format

**rocminfo**:

* Uses lower-level HSA (Heterogeneous System Architecture) interfaces
* Provides more comprehensive system-level hardware details
* Shows additional information about CPU agents, memory pools, and caches
* Reports hardware topology and NUMA relationships
* Includes ISA details and agent characteristics

> Use ``hipInfo`` when you need to quickly assess GPU compute capabilities and memory, check device capabilities for HIP programming decisions, and verify device properties available to your HIP applications. 

Reviewer 2: ```suggestion
Use ``hipInfo`` to quickly assess GPU compute capabilities and memory, check device capabilities for HIP programming decisions, and verify the device properties available to your HIP applications. 
```
